File: examples/typegraphs/microservice-orchestration.py
# skip:start
from os import environ

from typegraph import typegraph, Policy, t, Graph
from typegraph.graph.params import Cors
from typegraph.runtimes import DenoRuntime, GraphQLRuntime


# skip:end
@typegraph(
    # skip:next-line
    cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
)
def team_a(g: Graph):
    public = Policy.public()

    deno = DenoRuntime()
    records = GraphQLRuntime(environ.get("TG_URL", "http://localhost:7890") + "/team-b")

    g.expose(
        public,
        version_team_b=records.query(t.struct({}), t.integer(), path=["version"]),
        version_team_a=deno.static(t.integer(), 3),
    )


# @typegraph(
#     # skip:next-line
#     cors=Cors(allow_origin=["https://metatype.dev", "http://localhost:3000"]),
# )
# def team_b(g: Graph):
#     public = Policy.public()
#
#     deno = DenoRuntime()
#
#     g.expose(
#         public,
#         version=deno.static(t.integer(), 12),
#         record=deno.static(t.struct({"weight": t.integer()}), {"weight": 100}),
#     )

File: examples/typegraphs/microservice-orchestration.ts
// skip:start

import { Policy, t, typegraph } from "@typegraph/sdk/index.ts";
import { DenoRuntime } from "@typegraph/sdk/runtimes/deno.ts";
import { GraphQLRuntime } from "@typegraph/sdk/runtimes/graphql.ts";

/* eslint-disable  @typescript-eslint/no-explicit-any */

function getEnvOrDefault(key: string, defaultValue: string) {
  const glob = globalThis as any;
  const value = glob?.process
    ? glob?.process.env?.[key]
    : glob?.Deno.env.get(key);
  return value ?? defaultValue;
}
// skip:end

typegraph(
  {
    name: "team-a",
    // skip:next-line
    cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
  },
  (g) => {
    const pub = Policy.public();

    const deno = new DenoRuntime();
    const records = new GraphQLRuntime(
      getEnvOrDefault("TG_URL", "http://localhost:7890" + "/team-b")
    );

    g.expose(
      {
        version_team_b: records.query(t.struct({}), t.integer(), ["version"]),
        version_team_a: deno.static(t.integer(), 3),
      },
      pub
    );
  }
);

typegraph(
  {
    name: "team-b",
    // skip:next-line
    cors: { allowOrigin: ["https://metatype.dev", "http://localhost:3000"] },
  },
  (g) => {
    const pub = Policy.public();

    const deno = new DenoRuntime();

    g.expose(
      {
        version: deno.static(t.integer(), 12),
        record: deno.static(t.struct({ weight: t.integer() }), { weight: 100 }),
      },
      pub
    );
  }
);
